package NYT.NObjectClient.NProto;

import "yt_proto/yt/core/misc/proto/guid.proto";

////////////////////////////////////////////////////////////////////////////////

message TReqExecute
{
    // For forwarded and redirected requests, contains the original request id.
    optional NYT.NProto.TGuid original_request_id = 6;

    // Each Execute call carries a number of YPath requests.
    // Each of these requests is represented by a message.
    // Each message, in turn, is a sequence of parts.
    // This array contains the lengths of those part sequences.
    // In particular, its length is just the number of encapsulated YPath requests.
    repeated int32 part_counts = 1;

    // If |true|, the master is allowed to send a partial response to this request.
    optional bool allow_backoff = 3 [default = false];

    // If |true|, the client supports TRspExecute.subresponses and TRspExecute.uncertain_subrequest_indexes.
    optional bool supports_portals = 5;

    // Current sticky group size for batch
    optional int32 current_sticky_group_size = 8;

    // COMPAT(shakurov): deprecated; use TMulticellSyncExt instead.
    optional bool suppress_upstream_sync = 2 [default = false];
    optional bool suppress_transaction_coordinator_sync = 7 [default = false];

    // When Cypress Proxy is used, requests to different cells and peer kinds
    // are sent to the same proxies. These fields are used by Cypress Proxy to
    // redirect request properly. Master ignores these fields.
    optional uint32 cell_tag = 9;
    optional int32 master_channel_kind = 10; // EMasterChannelKind
}

message TRspExecute
{
    message TSubresponse
    {
        // The index of the subrequest.
        required int32 index = 1;
        // The number of parts in this subresponse.
        required int32 part_count = 2;
        // Hydra revision at the beginning of the subrequest execution.
        required uint64 revision = 3;
        // Optimal sticky group size depending on request rate and response size.
        optional int32 advised_sticky_group_size = 4;
    }
    // If non-empty, gives the detailed subresponse information in the new format.
    repeated TSubresponse subresponses = 3;

    // The indexes of subrequests that were forwarded to another master but
    // for which no response was received; the client must annotate the subsequent attempts with
    // "retry" flag.
    repeated int32 uncertain_subrequest_indexes = 4;

    // If true, next level cache is enabled.
    optional bool caching_enabled = 5 [default = false];

    // COMPAT(babenko): this is a legacy format; drop it
    // Similar to TReqExecute::part_counts, but denotes the number of parts in each
    // encapsulated response.
    // NB: A batched response may have less subresponses than its corresponding request
    // does subrequests if TReqExecute::allow_backoff was |true| (because of timeouts).
    repeated int32 part_counts = 1;

    // COMPAT(babenko): this is a legacy format; drop it
    // Hydra revision at the beginning of each YPath request execution.
    repeated uint64 revisions = 2;

    // COMPAT(babenko): this is a legacy format; drop it
    // Optimal sticky group size depending on request rate and response size.
    repeated int32 advised_sticky_group_size = 6;
}

////////////////////////////////////////////////////////////////////////////////

message TReqGCCollect
{ }

message TRspGCCollect
{ }

////////////////////////////////////////////////////////////////////////////////
